app/src/main/java/com/postpc/tenq/ui/activities/JoinQrActivity.java

```diff
@@ -1,13 +1,144 @@
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+public class JoinQrActivity extends TenQActivity {
 
-public class JoinQrActivity extends AppCompatActivity {
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
-        setContentView(R.layout.activity_join_qr);
+        IntentIntegrator intentIntegrator = new IntentIntegrator(this);
+        intentIntegrator.setPrompt("");
+        intentIntegrator.setOrientationLocked(true);
+        intentIntegrator.initiateScan();
+
+    }
+
+    /**
+     * on result of QR scanner
+     */
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
+        super.onActivityResult(requestCode, resultCode, data);
+        IntentResult intentResult = IntentIntegrator.parseActivityResult(requestCode, resultCode, data);
+        if (intentResult != null) {
+            if (intentResult.getContents() == null) {
+                Toast.makeText(getBaseContext(), "Cancelled", Toast.LENGTH_SHORT).show();
+            } else {
+                // string of QR code result
+                String qrResult = intentResult.getContents();
+
+                // update fire-store
+                FirebaseFirestore.getInstance().collection("rooms").document(qrResult)
+                        .get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
+                    @Override
+                    public void onComplete(@NonNull @NotNull Task<DocumentSnapshot> task) {
+                        if (task.isSuccessful()) {
+                            DocumentSnapshot document = task.getResult();
+                            if (document.exists()) {
+                                Room room = document.toObject(Room.class);
+                                // check if room is active
+                                if (room.isActive()) {
+                                    activeRoomFlow(room);
+                                } else {
+                                    Log.d("JoinQrActivity", "No such document");
+                                }
+                            } else {
+                                Log.d("JoinQrActivity", "get failed with ", task.getException());
+                            }
+                        }
+                    }
+
+                    private void activeRoomFlow(Room room) {
+                        Intent intent = new Intent(JoinQrActivity.this, RoomActivity.class);
+                        intent.putExtra("room_id", qrResult);
+
+                        // 1. add user id to guests list of room in fire-store (if not there already)
+                        addUserToRoomGuests(room);
+                        // 2. add room id to user rooms in fire-store (if not there already)
+                        addRoomToUserRooms(room);
+
+                        startActivity(intent);
+                        finish();
```

Comment:
From this point there's really a lot of duplicate code.
Wanna try extracting it to a ViewModel for both activities? We can do it together, I just studied it and used it for the RoomActivity